repose-aggregator/tests/functional-tests/src/integrationTest/groovy/features/filters/valkyrie/ValkyrieAuthorizationQualityTest.groovy

```diff
@@ -0,0 +1,176 @@
+
+        then:
+        mc.handlings.size() == 1
+        mc.handlings[0].request.path == "/"
+        List<String> tenantIdValues = mc.handlings[0].request.headers.findAll(TENANT_ID)
+        tenantIdValues.contains("hybrid:$contactId$quality" as String)
+        mc.receivedResponse.code as Integer == SC_OK
+
+        where:
+        [a, b, c] << [["GET", "HEAD", "PUT", "POST", "DELETE", "PATCH"],
+                      [["edit_product", 1234],
+                       ["admin_product", 2345],
+                       ["account_admin", 3456]],
+                      [";q=0.98", ""]
+        ].combinations()
+
+        method = a
+        devicePermission = b[0]
+        contactId = b[1]
+        quality = c
+
+    }
+
+    @Unroll
+    def "should allow #method calls with permission 'view_product' and quality '#quality'"() {
+        given:
+        def authToken = UUID.randomUUID().toString()
+        mockValkyrie.with {
+            valid_auth = authToken
+            contact_id = "4567"
+            device_perm = "view_product"
+        }
+
+
+        when:
+        MessageChain mc = deproxy.makeRequest(
+            url: "$reposeEndpoint/",
+            method: method,
+            headers: [
+                (AUTH_TOKEN)   : authToken,
+                (TENANT_ID)    : "someOtherTenant;q=0.5,hybrid:4567$quality",
+                (CONTACT_ID)   : mockValkyrie.contact_id,
+                (HDR_DEVICE_ID): mockValkyrie.device_id])
+
+        then:
+        mc.handlings.size() == 1
+        mc.handlings[0].request.path == "/"
+        List<String> tenantIdValues = mc.handlings[0].request.headers.findAll(TENANT_ID)
+        tenantIdValues.contains("hybrid:4567$quality" as String)
+        mc.receivedResponse.code as Integer == SC_OK
+
+        where:
+        [a, b] << [["GET", "HEAD"],
+                   [";q=0.98", ""]
+        ].combinations()
+
+        method = a
+        quality = b
+
+    }
+
+    @Unroll
+    def "should not allow #method calls with permission 'view_product' and quality '#quality'"() {
+        given:
+        def authToken = UUID.randomUUID().toString()
+        mockValkyrie.with {
+            valid_auth = authToken
+            contact_id = "4567"
+            device_perm = "view_product"
+        }
+
+
+        when:
+        MessageChain mc = deproxy.makeRequest(
+            url: "$reposeEndpoint/",
+            method: method,
+            headers: [
+                (AUTH_TOKEN)   : authToken,
+                (TENANT_ID)    : "someOtherTenant;q=0.5,hybrid:4567$quality",
+                (CONTACT_ID)   : mockValkyrie.contact_id,
+                (HDR_DEVICE_ID): mockValkyrie.device_id])
+
+        then:
+        mc.handlings.size() == 0
+        mc.receivedResponse.code as Integer == SC_FORBIDDEN
+
+        where:
+        [a, b] << [["PUT", "POST", "DELETE", "PATCH"],
+                   [";q=0.98", ""]
+        ].combinations()
+
+        method = a
+        quality = b
+
+    }
+}
```

Comment:
I would love to see a test that "fails" (passes, but with `SC_FORBIDDEN`) because the wrong tenant ID is used (as opposed to "failing" because the permission does not grant authorization). That would probably involve writing a replacement Valkyrie handler that would return a the failure template on one tenant ID, and the success template on another tenant ID. Then you can combine templates and qualities and make sure that you get back the right template (indirectly, via inspection of the response) for each request.

Reply:
why would that happen? all we do is get device permissions and make decisions on that?

Reply:
Right, so if the wrong tenant ID value was selected from the header, we might get an error response back from Valkyrie? Because the tenant we were requesting permissions for is not in Valkyrie or whatever the case may be. Basically, I just want a case capturing what happens when the wrong tenant ID is selected.

Reply:
pretty sure there are already existing tests for that, from before this change.